src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,95 @@
+package duke;
+
+import duke.commands.Command;
+import duke.exceptions.*;
+import duke.parser.Parser;
+import duke.storage.Storage;
+import duke.tasks.TaskList;
+import duke.ui.Ui;
+
+import java.util.Scanner;
+
+/**
+ * The main class for the Ding Chatbot application.
+ */
+public class Duke {
+
+    private TaskList tasks;
+    private Storage storage;
+
+    /**
+     * Constructs a new instance of the duke.Duke class.
+     * Initializes the storage and task list.
+     */
+    public Duke() {
+        storage = new Storage("./data", "./data/ding.txt");
+        tasks = storage.getTasks();
+    }
+
+    /**
+     * Runs the main loop of the chatbot application.
+     * Handles user input, executes commands, and interacts with the user.
+     */
+    public void run() {
+        Ui.introReply();
+        Scanner sc = new Scanner(System.in);
+        String str = sc.nextLine();
+        boolean isRunning = true;
+        while (isRunning) {
+            try {
+                Command command = Parser.getCommand(str);
+                switch (command) {
+                case LIST:
+                    Ui.printListReply(tasks);
+                    break;
+                case MARK:
+                case UNMARK:
+                    Ui.markTaskReply(Parser.taskToMark(str, tasks), tasks, command);
+                    break;
+                case TODO:
+                case DEADLINE:
+                case EVENT:
+                    tasks.addTask(Parser.parseStringToTask(str));
+                    Ui.addTaskReply(str, tasks);
+                    break;
+                case DELETE:
+                    Ui.deleteTaskReply(Parser.taskToDelete(str, tasks), tasks);
+                    break;
+                case FIND:
+                    Ui.findTaskReply(str, Parser.findKeyword(str, tasks));
+                    break;
+                case BYE:
+                    isRunning = false;
+                    Ui.exitReply();
+                    return;
```

Comment:
Perhaps you could consider handling the different cases separately rather than using waterfall? Might be hard to trace the flow of the program, even though waterfall will reduce code duplication